src/twisted/application/runner/_exit.py

```diff
@@ -6,35 +6,9 @@
 System exit support.
 """
 
+from enum import IntEnum
 from sys import stdout, stderr, exit as sysexit
-
-from constantly import Values, ValueConstant
-
-
-def exit(status, message=None):
-    """
-    Exit the python interpreter with the given status and an optional message.
-
-    @param status: An exit status.
-    @type status: L{int} or L{ValueConstant} from L{ExitStatus}.
-
-    @param message: An options message to print.
-    @type status: L{str}
-    """
-    if isinstance(status, ValueConstant):
-        code = status.value
-    else:
-        code = int(status)
-
-    if message:
-        if code == 0:
-            out = stdout
-        else:
-            out = stderr
-        out.write(message)
-        out.write("\n")
-
-    sysexit(code)
+from typing import NoReturn, Optional, Union
```

Comment:
[`typing.NoReturn`](https://docs.python.org/3/library/typing.html#typing.NoReturn) was added in Python 3.5.4 and 3.6.2. Some options to deal with this:
- bump Twisted's version requirements slightly
- import `NoReturn` from `typing_extensions` instead and add that package as a runtime dependency

Or if it would be acceptable to require 3.5.4/3.6.2+ only during type checking:
- catch `ImportError` and fall back to `None` if it's not found
- `import typing` and use `"typing.NoReturn"` (annotation as string)

Reply:
If we want to use `Protocol` or `Literal` in the near future, `typing_extensions` would be the best solution.

Reply:
Don't bother with `ImportError` - do an `if TYPE_CHECKING:` and only do the import in there. I'm totally fine with requiring more recent pythons for type checking.  I guess we have to live with quoted identifiers because 3.5 doesn't have `from __future__ import annotations` though?

Reply:
Even [`TYPE_CHECKING`](https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING) doesn't exist in 3.5.0 though; I learned today that it was added in 3.5.2.

A few weeks ago we had a 3.5.0 machine in the CI pool, but I don't know if that is still the case after the recent CI consolidation efforts.

`from __future__ import annotations` is available since 3.7, so it will be a while until we can use it, unless we decide to skip 3.6 and jump from 3.5 to 3.7 after the next release.

Reply:
3.5 is basically unused as I understand it, so I have no strong feelings about dropping it, but 3.6 is the most recent available pypy3 language version so that one's a bit of a non-starter :-(

Reply:
FWIW, all of our CI passed with this change…

Reply:
…which means that whatever change we make to accommodate this will not be confirmed to work in our CI.

Which to me implies that our support for Python versions that break in this way is not actually a thing. Right?

Reply:
Ah, no `pypy3` is seeing this, OK.  I'll go ahead and fix.

Reply:
Catching `ImportError` and falling back to `None` is problematic because `mypy` doesn't like `None`-or-`NoReturn` as a type. The forward reference works.